### PR TITLE
Add fancy History RAM API method to allow fetching multiple History RAM samples in a single API call

### LINKED ISSUE #1071: Improve usability of nidigital History RAM API
* Description of issue
Currently, History RAM (HRAM) configuration can be done via a set of properties or using a set of `configure_history_ram_xyz` methods. Fetching of HRAM samples is done by calling `get_history_ram_sample_count` to get the number of samples and then calling `fetch_history_ram_cycle_xyz` in a loop.

Usability can be improved by getting rid of `configure_history_ram_xyz` methods and providing a multi-sample fetch API.

### PR DESCRIPTION
**Pull request by sbethur**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).

- [x] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.

- [x] I've added tests applicable for this pull request

### What does this Pull Request accomplish?

Provide multi-sample History RAM fetch API

TODO: List of issues.

* Fix #1071 

### What testing has been done?

Added new system tests and ran them on real h/w (See #1273 for more info)

### CODE REVIEW

File: src/nidigital/custom_types/history_ram_cycle_information.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            'vector_number={}'.format(self.vector_number),
+            'cycle_number={}'.format(self.cycle_number),
+            'scan_cycle_number={}'.format(self.scan_cycle_number),
+            'expected_pin_states={}'.format(self.expected_pin_states),
+            'actual_pin_states={}'.format(self.actual_pin_states),
+            'per_pin_pass_fail={}'.format(self.per_pin_pass_fail),
+        ]
+
+        return '{0}({1})'.format(self.__class__.__name__, ', '.join(parameter_list))
+        return self.__class__.__name__
```
Comment (texasaggie97) [inline src/nidigital/custom_types/history_ram_cycle_information.py]: The second return is redundant, I think.

File: src/nidigital/metadata/functions_addon.py [hunk 2]
```diff
@@ -142,3 +142,71 @@
...
+                'documentation': {
+                    'description': 'Site on which to retrieve History RAM data. Specify site as a string in the form of siteN,\nwhere N is the site number. The VI returns an error if more than one site is specified.'
+                },
+                'name': 'site',
+                'type': 'ViConstString'
+            },
+            {
+                'direction': 'in',
+                'documentation': {
+                    'description': 'Pins for which to retrieve History RAM data. If empty, the pin list from the pattern\ncontaining the start label is used. Call get_pattern_pin_list or get_pattern_pin_names with the start\nlabel to retrieve the pins associated with the pattern burst.'
```
Comment (texasaggie97) [inline src/nidigital/metadata/functions_addon.py]: Reference methods as `niDigital_BlahBlah` so they will be turned into a link in help.

File: src/nidigital/metadata/functions_addon.py [hunk 3]
```diff
@@ -142,3 +142,71 @@
...
+        'codegen_method': 'python-only',
+        'method_templates': [
+            {
+                'documentation_filename': 'default_method',
+                'method_python_name_suffix': '',
+                'session_filename': 'fancy_fetch_history_ram_cycle_information',
+            }
+        ],
+        'documentation': {
+            'description': '\nReturns the pattern information acquired for the specified cycles.\n\nIf the pattern is using the edge multiplier feature, cycle numbers represent tester cycles, each of which may\nconsist of multiple DUT cycles. When using pins with mixed edge multipliers, pins may return\nNIDIGITAL_VAL_PIN_STATE_NOT_ACQUIRED for DUT cycles where those pins do not have edges defined.\n'
```
Comment (texasaggie97) [inline src/nidigital/metadata/functions_addon.py]: Does `PIN_STATE_NOT_ACQUIRED` exist as an enum value somewhere? It looks like this isn't correctly being turned into a link, and that would be one reason it wouldn't be.

File: src/nidigital/metadata/functions_addon.py [hunk 4]
```diff
@@ -142,3 +142,71 @@
...
... set for the acquired cycle.\n-  **vector_number** (int) Vector number within the pattern for the acquired cycle. Vector numbers start\nat 0 from the beginning of the pattern.\n-  **cycle_number** (int) Cycle number acquired by this History RAM sample. Cycle numbers start at 0\nfrom the beginning of the pattern burst.\n-  **scan_cycle_number** (int) Scan cycle number acquired by this History RAM sample. Scan cycle numbers\nstart at 0 from the first cycle of the scan vector. Scan cycle numbers are -1 for cycles that do not\nhave a scan opcode.\n-  **expected_pin_states** (list) Pin state as expected by the loaded pattern in the order specified in\nthe pin list. Pins without defined edges in the specified DUT cycle will have a value of\nNIDIGITAL_VAL_PIN_STATE_NOT_ACQUIRED.\n-  **actual_pin_states** (list) Pin state acquired by History RAM in the order specified in the pin\nlist. Pins without defined edges in the specified DUT cycle will have a value of\nNIDIGITAL_VAL_PIN_STATE_NOT_ACQUIRED.\n-  **per_pin_pass_fail** (list) pass fail information for pins in the order specified in the pin list.\nPins without defined edges in the specified DUT cycle will have a value of pass (True).\n'
```
Comment (texasaggie97) [inline src/nidigital/metadata/functions_addon.py]: For very long descriptions, it may be easier to format it using the triple quote:
```python
'description': """
This is a very long description
that has multiple
lines breaks as well as
* a list
* because
* why not?
"""
```

File: src/nidigital/custom_types/history_ram_cycle_information.py (hunk 1 above)
Comment (sbethur) [reply src/nidigital/custom_types/history_ram_cycle_information.py]: Yeah, that was a placeholder until I implemented the method body. Forgot to delete it later.

File: src/nidigital/metadata/functions_addon.py (hunk 2 above)
Comment (sbethur) [reply src/nidigital/metadata/functions_addon.py]: Done

File: src/nidigital/metadata/functions_addon.py (hunk 3 above)
Comment (sbethur) [reply src/nidigital/metadata/functions_addon.py]: I'm working on another PR where I update to using enums everywhere. I initially thought of including this in that PR, but I don't really have to wait until that PR. I fixed it here.

File: src/nidigital/metadata/functions_addon.py (hunk 4 above)
Comment (sbethur) [reply src/nidigital/metadata/functions_addon.py]: This is much better. Thanks for the tip!

File: generated/nidigital/nidigital/history_ram_cycle_information.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        row_format_d = '{:<20}: {:,}\n'
+        row_format_s = '{:<20}: {:}\n'
+
+        string_representation = ''
+        string_representation += row_format_s.format('Pattern Name', self.pattern_name)
+        string_representation += row_format_s.format('Time Set Name', self.time_set_name)
+        string_representation += row_format_d.format('Vector Number', self.vector_number)
+        string_representation += row_format_d.format('Cycle Number', self.cycle_number)
+        string_representation += row_format_d.format('Scan Cycle Number', self.scan_cycle_number)
+        string_representation += row_format_s.format('Expected Pin States', self.expected_pin_states)
```
Comment (texasaggie97) [inline generated/nidigital/nidigital/history_ram_cycle_information.py]: You format pin states differently in `__repr__`, should you here as well?

File: generated/nidigital/nidigital/session.py [hunk 6]
```diff
@@ -2265,6 +2267,121 @@ def fetch_capture_waveform(self, site_list, waveform_name, samples_to_read, time
...
+
+        if samples_to_read < -1:
+            raise ValueError('samples_to_read should be greater than or equal to -1.')
+
+        samples_available = self.get_history_ram_sample_count(site)
+        if position >= samples_available:
+            raise ValueError('position: Specified value = {0}, Maximum value = {1}.'.format(position, samples_available - 1))
+
+        if samples_to_read == -1:
+            if not self.history_ram_number_of_samples_is_finite:
```
Comment (texasaggie97) [inline generated/nidigital/nidigital/session.py]: I may be (probably?) reading this wrong, but shouldn't the exception be when number of samples _is_ infinite? Not when it isn't?

File: generated/nidigital/nidigital/session.py [hunk 7]
```diff
@@ -2265,6 +2267,121 @@ def fetch_capture_waveform(self, site_list, waveform_name, samples_to_read, time
...
+                raise RuntimeError(
+                    'Specifying -1 to fetch all History RAM samples is not supported when the digital pattern instrument is '
+                    'configured for continuous History RAM acquisition. You must specify an exact number of samples to fetch.')
+            samples_to_read = samples_available - position
+
+        if position + samples_to_read > samples_available:
+            raise ValueError(
+                'position: Specified value = {0}, samples_to_read: Specified value = {1}; Samples available = {2}.'
```
Comment (texasaggie97) [inline generated/nidigital/nidigital/session.py]: You don't actually say why this is an exception. Maybe add something about requesting beyond the samples available? Not sure what the wording should be though.

File: src/nidigital/metadata/enums_addon.py [hunk 8]
```diff
@@ -2,5 +2,7 @@
 # Any changes to the API should be made here. enums.py is code generated
 
 enums_override_metadata = {
+    'DigitalState': {
+        'codegen_method': 'public',
```
Comment (texasaggie97) [inline src/nidigital/metadata/enums_addon.py]: Add a comment why you need to set this.

File: generated/nidigital/nidigital/history_ram_cycle_information.py (hunk 5 above)
Comment (sbethur) [reply generated/nidigital/nidigital/history_ram_cycle_information.py]: I was debating between verbose info for pin states in `__str__` i.e. `<DigitalState.D: 8>` vs. just `DigitalState.D`. Thinking more about it, I don't see any value in the verbose way. I changed it to match `__repr__`.

File: generated/nidigital/nidigital/session.py (hunk 6 above)
Comment (sbethur) [reply generated/nidigital/nidigital/session.py]: The name of `history_ram_number_of_samples_is_finite` is confusing. If it is set to `false`, HRAM streaming mode is enabled. In this mode, users are not allowed to pass -1 to `samples_to_read` (which is supposed to mean "all available samples").
The code is correct.

File: generated/nidigital/nidigital/session.py (hunk 7 above)
Comment (sbethur) [reply generated/nidigital/nidigital/session.py]: The error description includes the number of samples user requested and number of samples available. This plus the fact that type of exception is `ValueError` should be enough information to let the user know what's wrong. I'm trying not to deviate too much from other nidigital API when it comes to error description.

Looking closely at this error message revealed a bug in the value calculated within the error message :)

File: src/nidigital/metadata/enums_addon.py (hunk 8 above)
Comment (sbethur) [reply src/nidigital/metadata/enums_addon.py]: done

File: generated/nidigital/nidigital/session.py (hunk 7 above)
Comment (texasaggie97) [reply generated/nidigital/nidigital/session.py]: But you have to already know the relationship between the values to know that `position + samples_to_read > samples_available` is the error condition. If you get a ValueError with the description "position: Specified value = 5, samples_to_read: Specified value = 100; Samples available = 40." there is nothing to say _why_ that is bad.

File: src/nidigital/templates/session.py/fancy_fetch_history_ram_cycle_information.py.mako [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            if pattern_index not in pattern_names:
+                pattern_names[pattern_index] = self.get_pattern_name(pattern_index)
+            pattern_name = pattern_names[pattern_index]
+
+            if time_set_index not in time_set_names:
+                time_set_names[time_set_index] = self.get_time_set_name(time_set_index)
+            time_set_name = time_set_names[time_set_index]
+
+            vector_number = vector_number
+            cycle_number = cycle_number
```
Comment (texasaggie97) [inline src/nidigital/templates/session.py/fancy_fetch_history_ram_cycle_information.py.mako]: Why are you setting these equal to themselves?

Comment (sbethur) [reply src/nidigital/templates/session.py/fancy_fetch_history_ram_cycle_information.py.mako]: No reason :). Forgot to delete this after renaming and restructuring some code. Cleaned them up.

File: docs/nidigital/class.rst [hunk 10]
```diff
@@ -1188,72 +1188,13 @@ fetch_history_ram_cycle_information
 
     .. py:currentmodule:: nidigital.Session
 
-    .. py:method:: fetch_history_ram_cycle_information(site, sample_index)
+    .. py:method:: fetch_history_ram_cycle_information(site, pin_list, position, samples_to_read)
```
Comment (marcoskirsch) [inline docs/nidigital/class.rst]: with sites becoming a repeated capability ( #1290 ) we will need to integrate this carefully.

File: docs/nidigital/class.rst [hunk 11]
```diff
@@ -1262,87 +1203,74 @@ fetch_history_ram_cycle_pin_data
...
-
-
-                per_pin_pass_fail (list of bool): 
-
-
-                    
-
-
-
-fetch_history_ram_scan_cycle_number
```
Comment (marcoskirsch) [inline docs/nidigital/class.rst]: a bunch of methods went away, consider mentioning that in CHANGELOG.md. It can still be a single "changed" item.

File: docs/nidigital/class.rst [hunk 12]
```diff
@@ -1262,87 +1203,74 @@ fetch_history_ram_cycle_pin_data
...
+                Number of samples to fetch. A value of -1 specifies to fetch all available samples.
 
                 
 
 
-            :type sample_index: int
+            :type samples_to_read: int
 
-            :rtype: int
+            :rtype: list of HistoryRAMCycleInformation
```
Comment (marcoskirsch) [inline docs/nidigital/class.rst]: Is it worth creating a type for this?
Why not return a list of dictionaries?

(consistency with `niscope`?)

File: generated/nidigital/nidigital/_library.py [hunk 13]
```diff
@@ -6,6 +6,8 @@
 
 from nidigital._visatype import *  # noqa: F403,H303
 
+import nidigital.history_ram_cycle_information as history_ram_cycle_information  # noqa: F401
+
```
Comment (marcoskirsch) [inline generated/nidigital/nidigital/_library.py]: why is this needed by `_library.py`?

File: generated/nidigital/nidigital/session.py [hunk 14]
```diff
@@ -2267,6 +2269,119 @@ def fetch_capture_waveform(self, site_list, waveform_name, samples_to_read, time
...
+                   Length of the inner list will be equal to the number of pins requested.
+
+        '''
+        if position < 0:
+            raise ValueError('position should be greater than or equal to 0.')
+
+        if samples_to_read < -1:
+            raise ValueError('samples_to_read should be greater than or equal to -1.')
+
+        samples_available = self.get_history_ram_sample_count(site)
```
Comment (marcoskirsch) [inline generated/nidigital/nidigital/session.py]: does `self.get_history_ram_sample_count` need to be public?

File: generated/nidigital/nidigital/session.py [hunk 15]
```diff
@@ -2267,6 +2269,119 @@ def fetch_capture_waveform(self, site_list, waveform_name, samples_to_read, time
...
+        if position >= samples_available:
+            raise ValueError('position: Specified value = {0}, Maximum value = {1}.'.format(position, samples_available - 1))
```
Comment (marcoskirsch) [inline generated/nidigital/nidigital/session.py]: There's a bunch of parameter value checks here. What's the harm in letting the runtime handle it instead?

File: src/nidigital/metadata/enums_addon.py [hunk 16]
```diff
@@ -2,5 +2,10 @@
 # Any changes to the API should be made here. enums.py is code generated
 
 enums_override_metadata = {
+    # Codegen deduces the enum types to generate based on usage of enum types in public metadata.
```
Comment (marcoskirsch) [inline src/nidigital/metadata/enums_addon.py]: Do we intend for these things to remain in `_addon.py` files in the long term?

File: src/nidigital/nidigital.mak [hunk 17]
```diff
@@ -8,5 +8,9 @@ MODULE_FILES_TO_COPY := $(DEFAULT_PY_FILES_TO_COPY)
 
 RST_FILES_TO_GENERATE := $(DEFAULT_RST_FILES_TO_GENERATE)
 
+# Add custom types to copy
```
Comment (marcoskirsch) [inline src/nidigital/nidigital.mak]: Comment is not very useful. More useful would be to say why this needs to be copied and/or where to.
I assume you copied this from `niscope` so whatever you do do it in both locations - remove / leave alone / reword

File: docs/nidigital/class.rst (hunk 12 above)
Comment (marcoskirsch) [reply docs/nidigital/class.rst]: Ok, there's the `__repr__` stuff too. Resolving.

File: docs/nidigital/class.rst (hunk 10 above)
Comment (sbethur) [reply docs/nidigital/class.rst]: Yes, whichever PR goes last will merge the other's changes.

File: docs/nidigital/class.rst (hunk 11 above)
Comment (sbethur) [reply docs/nidigital/class.rst]: Had already update the changelog; added another item to better convey the changes.

File: generated/nidigital/nidigital/session.py (hunk 14 above)
Comment (sbethur) [reply generated/nidigital/nidigital/session.py]: Yes, there are multiple use-cases where users may want to know the total count of HRAM samples. e.g. In HRAM streaming, users just want to know the number of failures in the pattern burst, but not necessarily fetch the HRAM data.

`get_history_ram_sample_count` is also in LV and .NET APIs.

File: generated/nidigital/nidigital/session.py (hunk 15 above)
Comment (sbethur) [reply generated/nidigital/nidigital/session.py]: Runtime doesn't do these checks since it doesn't support multi-sample fetch; it doesn't have any notion of `position` or `samples_to_read`.

These same checks are present in other APIs that support multi-sample fetch (LV and .NET), which is not good :(

File: src/nidigital/metadata/enums_addon.py (hunk 16 above)
Comment (sbethur) [reply src/nidigital/metadata/enums_addon.py]: No, it will go away when #1066 is fixed.

File: src/nidigital/nidigital.mak (hunk 17 above)
Comment (sbethur) [reply src/nidigital/nidigital.mak]: I removed it. None of the other rules in the file have any comments.

File: generated/nidigital/nidigital/_library.py (hunk 13 above)
Comment (sbethur) [reply generated/nidigital/nidigital/_library.py]: In this case it's not needed. But in the case of other custom types (like `waveform_info` in niscope) the module also contains a type that's passed to the C function. Unconditionally adding the import keeps the codegen and/or metadata simpler.
